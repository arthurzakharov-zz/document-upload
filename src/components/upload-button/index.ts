export { default } from "./upload-button.component";
