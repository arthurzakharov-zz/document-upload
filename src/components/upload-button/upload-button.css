.upload-button {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  box-shadow: 0 4px 10px var(--grey-5);
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--orange-1);
  cursor: pointer;
}

.upload-button--loaded {
  cursor: not-allowed;
  background-color: var(--white-1);
}

.upload-button__check {
  height: 15px;
  stroke: var(--green-2);
}

.upload-button__plus {
  height: 13px;
  fill: var(--white-1);
}

.upload-button__upload {
  height: 14px;
  fill: var(--white-1);
}

.upload-button__text {
  display: none;
  margin-right: 8px;
  font-family: "Titillium Web", sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 15px;
  line-height: 20px;
  color: var(--white-1);
}

@media screen and (min-width: 768px) {
  .upload-button {
    width: auto;
    padding: 10px 13px;
    border-radius: 20px;
  }

  .upload-button__text {
    display: block;
  }
}
