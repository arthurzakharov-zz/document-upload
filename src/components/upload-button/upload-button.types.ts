import type { MouseEvent } from "react";

export type UploadButtonPropsType = {
  loaded: boolean;
  multiple: boolean;
  onClick: (e: MouseEvent<HTMLButtonElement>) => void;
};
