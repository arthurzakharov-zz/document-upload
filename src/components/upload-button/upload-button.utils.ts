import cn from "classnames";

export const uploadButton = (loaded: boolean): string =>
  cn("upload-button", {
    "upload-button--loaded": loaded,
  });
