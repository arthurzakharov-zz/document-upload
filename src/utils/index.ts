import { DataSizeType } from "../types";

export const isObjectEmpty = (obj: Object): boolean => {
  // eslint-disable-next-line no-restricted-syntax
  for (const prop in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, prop)) {
      return false;
    }
  }
  return JSON.stringify(obj) === JSON.stringify({});
};

export const get = (obj: any, ...path: string[]): any => {
  return obj && path.reduce((result, p) => (result == null ? undefined : result[p]), obj);
};

export const convertDataSize = (number: number, from: DataSizeType = "B", to: DataSizeType = "MB"): number => {
  if (from === to) return number;
  const sizes: DataSizeType[] = ["B", "kB", "MB"];
  const fromIndex = sizes.indexOf(from);
  const toIndex = sizes.indexOf(to);
  const diff = fromIndex - toIndex;
  return diff < 0
    ? Math.round((number / 1024 ** Math.abs(diff)) * 100) / 100
    : Math.round(number * 1024 ** Math.abs(diff) * 100) / 100;
};

export const getUrlParam = (paramKey: string): string => {
  const urlSearchParams = new URLSearchParams(document.location.search);
  const paramKeyValues = urlSearchParams.getAll(paramKey);
  return paramKeyValues[paramKeyValues.length - 1] || "";
};
