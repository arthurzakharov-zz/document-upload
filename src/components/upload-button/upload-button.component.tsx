import { MouseEvent } from "react";
import SvgCheck from "../../svg/Check";
import SvgPlus from "../../svg/Plus";
import SvgUpload from "../../svg/Upload";
import { uploadButton } from "./upload-button.utils";
import "./upload-button.css";

export interface UploadButtonProps {
  loaded: boolean;
  multiple: boolean;
  onClick: (e: MouseEvent<HTMLButtonElement>) => void;
}

function UploadButton(props: UploadButtonProps) {
  const { loaded, multiple, onClick } = props;

  const onClickHandler = (e: MouseEvent<HTMLButtonElement>) => {
    if (!loaded) {
      onClick(e);
    }
  };

  return loaded ? (
    <div data-testid="upload-div" className={uploadButton(loaded)}>
      <SvgCheck data-testid="upload-button-check" className="upload-button__check" />
    </div>
  ) : (
    <button
      data-testid="upload-button"
      aria-label={multiple ? "upload-single-file" : "upload-multiple-file"}
      type="button"
      className={uploadButton(loaded)}
      onClick={onClickHandler}
    >
      <span className="upload-button__text">{multiple ? "Hinzuf√ºgen" : "Hochladen"}</span>
      {multiple ? (
        <SvgPlus data-testid="upload-button-plus" className="upload-button__plus" />
      ) : (
        <SvgUpload data-testid="upload-button-upload" className="upload-button__upload" />
      )}
    </button>
  );
}

export default UploadButton;
