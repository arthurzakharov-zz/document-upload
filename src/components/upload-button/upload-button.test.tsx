import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { uploadButton } from "./upload-button.utils";
import UploadButton from "./upload-button.component";

const UPLOAD_BUTTON = "upload-button";
const UPLOAD_BUTTON_UPLOAD = "upload-button-upload";
const UPLOAD_BUTTON_CHECK = "upload-button-check";

describe("UploadButton", () => {
  test("uploadButton is not loaded", () => {
    expect(uploadButton(false)).toEqual("upload-button");
  });
  test("uploadButton is loaded", () => {
    expect(uploadButton(true)).toEqual("upload-button upload-button--loaded");
  });
  test("Snapshot is in not loaded state", () => {
    const mockFn = jest.fn();
    const { container } = render(<UploadButton loaded={false} multiple onClick={mockFn} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test("Snapshot is in loaded state", () => {
    const mockFn = jest.fn();
    const { container } = render(<UploadButton loaded multiple onClick={mockFn} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test("In not loaded state upload icon and text are displayed", () => {
    const mockFn = jest.fn();
    render(<UploadButton loaded={false} multiple={false} onClick={mockFn} />);
    expect(screen.getByTestId(UPLOAD_BUTTON)).toHaveTextContent("Hochladen");
    expect(screen.getByTestId(UPLOAD_BUTTON_UPLOAD)).toBeInTheDocument();
  });
  test("In loaded state check icon is diplayed", () => {
    const mockFn = jest.fn();
    render(<UploadButton loaded multiple onClick={mockFn} />);
    expect(screen.getByTestId(UPLOAD_BUTTON_CHECK)).toBeInTheDocument();
  });
  test("Focus on button and press enter", () => {
    const mockFn = jest.fn();
    render(<UploadButton loaded={false} multiple onClick={mockFn} />);
    expect(screen.getByTestId(UPLOAD_BUTTON)).not.toHaveFocus();
    userEvent.tab();
    expect(screen.getByTestId(UPLOAD_BUTTON)).toHaveFocus();
    userEvent.keyboard("{Space}");
    userEvent.keyboard("{Enter}");
    expect(mockFn).toBeCalledTimes(1);
  });
  test("Click on button", () => {
    const mockFn = jest.fn();
    render(<UploadButton loaded={false} multiple onClick={mockFn} />);
    userEvent.click(screen.getByTestId(UPLOAD_BUTTON));
    expect(mockFn).toBeCalledTimes(1);
  });
});
