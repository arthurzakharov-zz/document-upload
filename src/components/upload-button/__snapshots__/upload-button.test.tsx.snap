// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadButton Snapshot is in loaded state 1`] = `
<div
  class="upload-button upload-button--loaded"
  data-testid="upload-div"
>
  <svg
    class="upload-button__check"
    data-testid="upload-button-check"
    fill="none"
    stroke="#2ecc71"
    viewBox="0 0 15 15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m2 7 4.213 7 6.62-12"
      stroke-linecap="square"
      stroke-linejoin="bevel"
      stroke-width="2"
    />
  </svg>
</div>
`;

exports[`UploadButton Snapshot is in not loaded state 1`] = `
<button
  aria-label="upload-single-file"
  class="upload-button"
  data-testid="upload-button"
  type="button"
>
  <span
    class="upload-button__text"
  >
    Hinzuf√ºgen
  </span>
  <svg
    class="upload-button__plus"
    data-testid="upload-button-plus"
    fill="#fff"
    viewBox="0 0 14 15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7.933 6.098H14v1.867H7.933v6.066H6.067V7.965L0 7.99V6.098l6.067-.026V.032h1.866v6.066Z"
    />
  </svg>
</button>
`;
